Wordcamp docker (#113)
WordPress/wordcamp.org: readme.md

### WordCamp.org

This repository mirrors https://meta.svn.wordpress.org/sites/trunk/wordcamp.org/ and can be used to contribute used Git instead of SVN.

###### Setup 

There are two primary ways to setup this repo for local development.

1. Using VVV. WordCamp.org is available as part of WordPress.org's meta network, and can be setup by following meta-environment setup guide: https://github.com/WordPress/meta-environment/blob/master/docs/install.md

> 1. If you'd like to only setup WordCamp.org and would like to use Docker, you can use inbuilt docker provisioning. Please follow these steps

Reviewer: `SHOULD` - I think the docker-specific stuff should be modularize the docker into `.docker/README.md`, and then this item should just link to that. That keeps this file more digestable, and focused on being an overview, and avoids overwhelming people who aren't using Docker with irrelevant info.

Rather than adding this file now, though, I think it'd be better to do that as part of #2, since that has a lot of other stuff we probably want to include.

Author: I would think Docker and VVV instructions should reside within same parent folder. Maybe same level pages on make/meta or inside a docs folder in git repo with readme linking to those. What do you think?

Reviewer: 🤔 Having them inside the repo is probably best, so that people don't have to switch back and forth between a browser and IDE/terminal. Having it in the repo also gives us a rendered version on GitHub.com that we can link to from the handbooks.

I think it'd be good for `/README.md` to _mention_ the VVV and Docker instructions, and link to them, but having the full instructions in the main readme feels overwhelming and cluttered to me. I think it'll make it difficult for people to navigate that document and find what they need.

The VVV instructions should probably remain inside the Meta Environment repo, though, since that's where the provisioning code/database/etc is. It'd be good to mention/link to it from `/README.md`, though.

I like the idea of a `docs/` folder, but it doesn't seem like we have enough for that yet, since right now it's just this file. Maybe as part of #2 we could discuss moving some of our internal devdocs there, though? Even if we did move it there, I still think the best location for the Docker readme would be inside `.docker/`, but it'd definitely be good to link (or just symlink) to that from `docs/`.

I'm open to other ideas, though. Do you have any thoughts, @coreymckrill ?

Author: I splitted readme.md in https://github.com/WordPress/wordcamp.org/pull/113/commits/db5db5a1d8c4db4cd3f90b0c97aa60795cfa729a . Please resolve this conversation if this is good enough for now.

Reviewer: 👍

    **Note:** This will create `.data` directory which will contain MySQL files to persist data across docker restarts.

    1. Make sure you have Docker installed and Docker daemon running on your system.
    
    1. Clone the repo: 
        ```
        git clone git@github.com:WordPress/wordcamp.org.git
        ```
    
    1. Change into the directory, and run docker compose command. This could take some time depending upon the internet speed.
        ```
        cd wordcamp.org && docker-compose up --build
        ```
         This will also start the docker environment for WordCamp.org development.
         
    1. Add the following resolutions to your host file:
        ```
        127.0.0.1	central.wordcamp.test
        127.0.0.1	wordcamp.test
        127.0.0.1	2014.content.wordcamp.test
        127.0.0.1	2014.misc.wordcamp.test
        127.0.0.1	2014.new-site.wordcamp.test
        ```
        
        `/wp-admin` pages for these sites should now be accessible. Use `admin` as username and `password` as password to login.
        

>         **Note:** `https` URL scheme must be used to visit these sites. Security exception will be required in first time run.

Reviewer: `MIGHT` - It'd be nice to add instructions for adding the cert to the browser, so you get a green link instead of yellow. That's not necessary, just a nice-to-have

Author: Punting this for later. Because the process will actually be other way around. You would generate a cert on host first and then add it inside Docker. 
Keeping this conversation un-resolved for now, and will open a issue when we merge this.

Reviewer: 👍

Author: Raised https://github.com/WordPress/wordcamp.org/issues/121

>         **Note:** `https` URL scheme must be used to visit these sites. Security exception will be required in first time run.

Reviewer: `MUST` - Do `HTTP` URLs redirect to `HTTPS`?

Author: Yes they do. See https://github.com/WordPress/wordcamp.org/pull/113/files#diff-821cdce799b2a90c846089e20e997ff2R8

Reviewer: 👍

>     1. (Optional) Basic installation comes without themes and with only very basic plugins. Add plugins and themes which are installed in wordcamp.org if needed like so:

Reviewer: `MUST` - Can this be automated as part of installation?

Author: Unfortunately I was not able to find a non complicated solution to do this. Main reason being that we mount the `wp-content` directory from host filesystem, which is going to overwrite Docker image's `wp-content` folder.

Reviewer: 🤔 I'm confused, I would think they'd be the same folder, so any changes within the container would be synced to the host. Is that not the case?

Author: Yup, so during the first time docker build process, a container image will be build. But while starting up the container, we will mount the host's `wp-content` directory inside that image (overwriting image's wp-content folder) so that files are synced. That's why if during the build process we install the plugins and themes, they will get overwritten. 

We can solve this by not mounting `wp-content` but instead sub-folders one by one, but that will make the config very big and hard to keep updated. Let me also take a look to see if there are other solutions, but so far this looks like to be the case.

Reviewer: 🤔 

huh. Maybe this is another case where we'd want [a script that the user runs on the host after installation](https://github.com/WordPress/wordcamp.org/pull/113#discussion_r284009060)?

Then it seems like we couldn't need to install anything during the initial build.

Author: We already have a script for installing plugins and themes in `install-pluign-theme.sh`. Perhaps we can mark this step as mandatory instead of optional. What do you think?

Reviewer: That sounds good to me, since they won't really have a completely functioning environment until they do that.

Author: Done in https://github.com/WordPress/wordcamp.org/pull/113/commits/8e679936b8e51e534ca63ea91ea26639384690de

    
        1. Find the docker which is running WordCamp. It will start with `wordcamporg_wordcamp.test` and will most likely be `wordcamporg_wordcamp.test_1`
        
        1. Go inside this docker:
        
           **Note:** You might need to change `wordcamporg_wordcamp.test_1` to whatever is docker name in your setup.
            ```bash
               docker exec -it wordcamporg_wordcamp.test_1 bash
            ```
        
        1. Run predefined command for installing plugins and themes. This could take some time depending on your internet connection:
            ```bash
               sh install-plugin-theme.sh
            ```
            
        1. Activate plugin or apply theme as needed.
        

>     1. (Optional) By default, docker will start with data defined in `.docker/wordcamp_dev.sql` and changes to data will be persisted across runs. To start with different database, delete `.data` directory and replace the `.docker/wordcamp_dev.sql` file and run `docker-compose up --build` again.

Reviewer: ```suggestion
    1. (Optional) By default, docker will start with data defined in `.docker/wordcamp_dev.sql` and changes to data will be persisted across runs in `.data/`. To start with different database, delete `.data` directory and replace the `.docker/wordcamp_dev.sql` file and run `docker-compose up --build` again.
```

Author: Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

    
    After first time provisioning, docker can be started by using `docker-compose up` command from inside the directory. 
        

> To contribute, you can send pull requests to this repo, or add patches to https://meta.trac.wordpress.org/.

Reviewer: ```suggestion
To contribute, you can send [pull requests on GitHub](https://github.com/WordPress/wordcamp.org/), or add [patches on Meta Trac](https://make.wordpress.org/meta/handbook/about/get-involved/learn-how-to-contribute-code/#share-your-new-code).
```

Author: Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8
